### PR TITLE
feat: add the test-cases for the ai package

### PR DESCRIPTION
**Pull request by Aisuko**
Closes # 

## 📑 Description

This PR adds the suit test for all the AI files. The suit test imports three new packages, `ginkgo/v2` and `gomega`, and `gomock`(to mock the interface to test the specific functions). 

And installing and regenerating mock files tool `mockgen` has been added to the `Makefile.

Also, all the test cases have passed but some cases still need more investigation, and I will keep working on some edge cases like solving the real HTTP request in `openAI` client repo always be a failed issue. I mean these cases are not give the maintainer 100% confidence currently.

`Source code changes:`
There are some source code changes, I believe `GetName` return to the pre-defined element is safer than the string.

## ✅ Checks

- [x] My pull request adheres to the code style of this project
- [ ] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [x] All the tests have passed

### CODE REVIEW

**Comment by Aisuko** [comment]
The coverage result for `github.com/k8sgpt-ai/k8sgpt/pkg/ai` below:
```
@Aisuko ➜ /workspaces/k8sgpt (feat/testcase) $ go test ./... -coverprofile cover.out
?       github.com/k8sgpt-ai/k8sgpt     [no test files]
?       github.com/k8sgpt-ai/k8sgpt/cmd [no test files]
?       github.com/k8sgpt-ai/k8sgpt/cmd/analyze [no test files]
?       github.com/k8sgpt-ai/k8sgpt/cmd/auth    [no test files]
?       github.com/k8sgpt-ai/k8sgpt/cmd/filters [no test files]
?       github.com/k8sgpt-ai/k8sgpt/cmd/generate        [no test files]
?       github.com/k8sgpt-ai/k8sgpt/cmd/integration     [no test files]
?       github.com/k8sgpt-ai/k8sgpt/cmd/serve   [no test files]
ok      github.com/k8sgpt-ai/k8sgpt/pkg/ai      0.020s  coverage: 52.8% of statements
ok      github.com/k8sgpt-ai/k8sgpt/pkg/analysis        0.032s  coverage: 9.1% of statements
?       github.com/k8sgpt-ai/k8sgpt/pkg/cache   [no test files]
?       github.com/k8sgpt-ai/k8sgpt/pkg/common  [no test files]
?       github.com/k8sgpt-ai/k8sgpt/pkg/integration     [no test files]
?       github.com/k8sgpt-ai/k8sgpt/pkg/integration/trivy       [no test files]
?       github.com/k8sgpt-ai/k8sgpt/pkg/kubernetes      [no test files]
?       github.com/k8sgpt-ai/k8sgpt/pkg/server  [no test files]
?       github.com/k8sgpt-ai/k8sgpt/pkg/util    [no test files]
ok      github.com/k8sgpt-ai/k8sgpt/pkg/analyzer        0.037s  coverage: 59.9% of statements
```

**Comment by AlexsJones** [comment]
Thank you so much for this wonderful contribution @Aisuko however, we are not using `ginkgo` in any of our tests and our purposefully using just the standard `testing` library. This is to make it easier for new folks to contribute and long-term maintenance. 

I understand if you don't want to do more work, but would you be happy to update these to use the standard testing library?

**Comment by Aisuko** [comment]
> Thank you so much for this wonderful contribution @Aisuko however, we are not using `ginkgo` in any of our tests and our purposefully using just the standard `testing` library. This is to make it easier for new folks to contribute and long-term maintenance.
> 
> I understand if you don't want to do more work, but would you be happy to update these to use the standard testing library?

No worries @AlexsJones, I will follow the existing testing cases format. By the way, would you mind inviting me to the Slack channel? Then I can post the plan first, and make sure to get you guys to confirm. To be honest, I'm not sure if it is ok I used `gomock` for the project.

**Comment by Aisuko** [comment]
This PR was not solved `openAi.client.CreateChatCompletion` always returns an error with the test cases, maybe need further PR with a test token to solve it.

```
@Aisuko ➜ /workspaces/k8sgpt (feat/testcase) $ go test ./... -coverprofile cover.out
?       github.com/k8sgpt-ai/k8sgpt     [no test files]
?       github.com/k8sgpt-ai/k8sgpt/cmd [no test files]
?       github.com/k8sgpt-ai/k8sgpt/cmd/analyze [no test files]
?       github.com/k8sgpt-ai/k8sgpt/cmd/auth    [no test files]
?       github.com/k8sgpt-ai/k8sgpt/cmd/filters [no test files]
?       github.com/k8sgpt-ai/k8sgpt/cmd/generate        [no test files]
?       github.com/k8sgpt-ai/k8sgpt/cmd/integration     [no test files]
?       github.com/k8sgpt-ai/k8sgpt/cmd/serve   [no test files]
ok      github.com/k8sgpt-ai/k8sgpt/pkg/ai      0.016s  coverage: 47.8% of statements
ok      github.com/k8sgpt-ai/k8sgpt/pkg/analysis        0.032s  coverage: 9.1% of statements
?       github.com/k8sgpt-ai/k8sgpt/pkg/cache   [no test files]
?       github.com/k8sgpt-ai/k8sgpt/pkg/common  [no test files]
?       github.com/k8sgpt-ai/k8sgpt/pkg/integration     [no test files]
?       github.com/k8sgpt-ai/k8sgpt/pkg/integration/trivy       [no test files]
?       github.com/k8sgpt-ai/k8sgpt/pkg/kubernetes      [no test files]
?       github.com/k8sgpt-ai/k8sgpt/pkg/server  [no test files]
?       github.com/k8sgpt-ai/k8sgpt/pkg/util    [no test files]
ok      github.com/k8sgpt-ai/k8sgpt/pkg/analyzer        0.036s  coverage: 59.9% of statements
```

**Comment by AlexsJones** [comment]
> > Thank you so much for this wonderful contribution @Aisuko however, we are not using `ginkgo` in any of our tests and our purposefully using just the standard `testing` library. This is to make it easier for new folks to contribute and long-term maintenance.
> > I understand if you don't want to do more work, but would you be happy to update these to use the standard testing library?
> 
> No worries @AlexsJones, I will follow the existing testing cases format. By the way, would you mind inviting me to the Slack channel? Then I can post the plan first, and make sure to get you guys to confirm. To be honest, I'm not sure if it is ok I used `gomock` for the project.


Thank you for the contribution https://join.slack.com/t/k8sgpt/shared_invite/zt-1rwe5fpzq-VNtJK8DmYbbm~iWL1H34nw is the slack invite! I will check it out now

**Comment by Aisuko** [comment]
Hi, @AlexsJones, can we merge this one?

File: pkg/ai/openai.go [hunk 1]
```diff
@@ -107,5 +107,5 @@ func (a *OpenAIClient) Parse(ctx context.Context, prompt []string, cache cache.I
 }
 
 func (a *OpenAIClient) GetName() string {
-	return "openai"
+	return Backends[0]
```
Comment (AlexsJones) [inline pkg/ai/openai.go:110]: Adding the indirection to the `Backends` slice doesn't really improve the readability of this , can we revert for all such occurrences in `GetName()` ?

File: pkg/ai/openai_test.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+	openAIClient := &OpenAIClient{}
+
+	err := openAIClient.Configure(mockConfig, "testLanguage")
+	if err != nil {
+		t.Errorf("Error configuring OpenAIClient: %v", err)
+	}
+
+	// @Aisuko - Need to mock c.client.CreateChatCompletion return value.
+	// Otherwise it will always throw error with test data
```
Comment (AlexsJones) [inline pkg/ai/openai_test.go:43]: Would it be helpful to add `GetCompletion` to the `IAI` interface so that we can mock this out?

**Review by AlexsJones — COMMENTED** [review]
It seems like these tests are being done before the improvement of the AI structures themselves which means the tests aren't yet really able to properly test anything. 

Would you like me to reimplement the AI providers so that they are more testable before we merge this in? Therefore you could then uncomment a lot of the code you have written?

**Comment by Aisuko** [comment]
> It seems like these tests are being done before the improvement of the AI structures themselves which means the tests aren't yet really able to properly test anything.
> 
> Would you like me to reimplement the AI providers so that they are more testable before we merge this in? Therefore you could then uncomment a lot of the code you have written?

Hi, @AlexsJones .  As you mentioned, it does not really test too much useful features because the message was sent using the external package `github.com/sashabaranov/go-openai `. We will look into how to solve this.

And if you are working on reimplementing AI providers, I believe we can hold this PR until you finish it. I want to make these test cases really helpful for the project.

File: pkg/ai/openai_test.go (hunk 2 above)
Comment (Aisuko) [reply pkg/ai/openai_test.go:43]: Yes, it would be more helpful on increasing the code coverage. The test cases cannot cover the functions that were imported from the `github.com/sashabaranov/go-openai` package. E.g: `c.client.CreateChatCompletion`

**Comment by Aisuko** [comment]
@AlexsJones Please let me know if we can work back on this PR.
